File: frontend/api/voice.js
import { API_BASE_URL } from '../utils/config'
import { post, upload } from './request'
import { splitToSentences } from '../utils/voice-reader'

const MOCK_AUDIO_URL = 'https://www.learningcontainer.com/wp-content/uploads/2020/02/Kalimba.mp3'

function getApiOrigin() {
  return API_BASE_URL.replace(/\/api\/v1$/, '')
}

function resolveAudioUrl(url) {
  if (!url) {
    return MOCK_AUDIO_URL
  }
  if (/^https?:\/\//.test(url)) {
    return url
  }
  return `${getApiOrigin()}${url}`
}

function mockChatResponse(message, context = '') {
  const text = `${context ? `我先结合当前页面内容说明一下。` : ''}您刚才说的是“${message}”。我会建议您一步一步来，如果看不清字，可以先点整页朗读，再根据提示点击对应入口。`
  return {
    text,
    audioUrl: MOCK_AUDIO_URL,
    provider: 'mock',
    sentences: splitToSentences(text)
  }
}

export async function chatWithVoiceAssistant(payload) {
  try {
    const response = await post('/voice/chat', payload)
    return {
      ...response.data,
      audioUrl: resolveAudioUrl(response.data.audioUrl)
    }
  } catch (error) {
    return mockChatResponse(payload.message, payload.context)
  }
}

export async function synthesizeText(text) {
  try {
    const response = await post('/voice/tts', { text })
    return {
      ...response.data,
      audioUrl: resolveAudioUrl(response.data.audioUrl)
    }
  } catch (error) {
    return {
      text,
      audioUrl: MOCK_AUDIO_URL,
      provider: 'mock',
      sentences: splitToSentences(text)
    }
  }
}

export async function readPageContent(title, sections) {
  try {
    const response = await post('/voice/page-read', { title, sections })
    return {
      ...response.data,
      audioUrl: resolveAudioUrl(response.data.audioUrl)
    }
  } catch (error) {
    const text = sections.join('\n')
    return {
      title,
      text,
      audioUrl: MOCK_AUDIO_URL,
      provider: 'mock',
      sentences: splitToSentences(text)
    }
  }
}

export async function transcribeVoice(filePath) {
  try {
    const response = await upload('/voice/asr', filePath, 'file')
    return response.data.text
  } catch (error) {
    return '帮我把当前页面读一遍。'
  }
}

File: backend/app/routers/voice.py
from fastapi import APIRouter, File, HTTPException, UploadFile

from app.core.config import settings
from app.schemas import VoiceChatIn, VoiceReadIn, VoiceReadOut, VoiceTtsIn
from app.services.voice import generate_chat_reply, save_uploaded_audio, split_sentences, synthesize_speech, transcribe_audio

router = APIRouter(prefix='/voice', tags=['voice'])


def success(data=None, message='success'):
  return {'code': 0, 'message': message, 'data': data}


@router.post('/chat')
def voice_chat(payload: VoiceChatIn):
  answer = generate_chat_reply(payload.message, payload.context or '')
  audio = synthesize_speech(answer)
  sentences = split_sentences(answer)
  return success({
    'text': answer,
    'audioUrl': audio['audioUrl'],
    'provider': audio['provider'],
    'sentences': sentences
  })


@router.post('/tts')
def voice_tts(payload: VoiceTtsIn):
  audio = synthesize_speech(payload.text)
  sentences = split_sentences(payload.text)
  return success({
    'text': payload.text,
    'audioUrl': audio['audioUrl'],
    'provider': audio['provider'],
    'sentences': sentences
  })


@router.post('/page-read', response_model=VoiceReadOut)
def voice_page_read(payload: VoiceReadIn):
  text = '\n'.join([item.strip() for item in payload.sections if item and item.strip()])
  audio = synthesize_speech(text)
  sentences = split_sentences(text)
  return VoiceReadOut(
    code=0,
    message='success',
    data={
      'title': payload.title,
      'text': text,
      'audioUrl': audio['audioUrl'],
      'provider': audio['provider'],
      'sentences': sentences
    }
  )


@router.post('/asr')
async def voice_asr(file: UploadFile = File(...)):
  file_bytes = await file.read()
  if not file_bytes:
    raise HTTPException(status_code=400, detail='empty file')
  saved = save_uploaded_audio(file_bytes, file.filename or 'voice.wav')
  text = ''
  if settings.public_file_base_url:
    try:
      public_url = f'{settings.public_file_base_url.rstrip("/")}{saved["relative_url"]}'
      text = transcribe_audio(public_url)
    except Exception:
      text = ''
  return success({
    'text': text or '当前环境还没有公网音频地址，真实语音识别暂时不可用。您可以先手动输入，或把后端部署到公网后再启用。'
  })
